_overviews/scala3-book/taste-contextual-abstractions.md

```diff
@@ -12,19 +12,27 @@ next-page: taste-toplevel-definitions
 TODO: Now that this is a separate section, it needs a little more content.
 {% endcomment %}
 
+Under certain circumstances, you can omit some parameters of method calls that are considered repetitive.
 
-Under certain circumstances, the Scala compiler can “write” some parts of your programs.
-For instance, consider a program that sorts a list of addresses by two criteria, city name and then street name:
+Those parameters are called _Context Parameters_ because there are "guessed" by the compiler from the context surrounding the method call.
+
+For instance, consider a program that sorts a list of addresses by two criteria: city name and then street name.
 
 ```scala
 val addresses: List[Address] = ...
-addresses.sortBy(address => (address.city, address.street))
+
+given ordering: Ordering[Address] = Ordering.by(address.city, address.street)
+
+addresses.sorted(using ordering)
+addresses.sorted
```

Comment:
I strongly believe that the previous example was really hard to understand for people that are not familiar with implicit parameters.

```scala
val addresses: List[Address] = ...
addresses.sortBy(address => (address.city, address.street))
```

Here we don't really now what a context is and what the compiler does exactly with it: passing a parameter to a method.

```scala
val addresses: List[Address] = ...
given ordering: Ordering[Address] = Ordering.by(address => address.city, address.street)

addresses.sorted(using ordering)
addresses.sorted
```

In my example I show how to alter the context with the `given` keyword, and how the compiler is able to pass it to a method, in this case `sorted`. It is more concrete and more easier to understand for beginners.

Reply:
I’m not convinced because you still use `Ordering.by` in the same way we used `sortBy` in the original example (so, the same critics apply to your suggestion).

Reply:
Okay so your point is that there is a context parameter that is given to the `Ordering.by` method but it is ignored in the explanation?

Reply:
Yes exactly :smile:

Reply:
What about this example?

```scala
val persons: List[Person] = ...

given byAge: Ordering[Person] = new Ordering[Person]:
  def compare(x: Person, y: Person) =
    if x.age < y.age then - 1 else if x.age > y.age then 1 else 0
  
persons.sorted(using byAge)
persons.sorted
```

Reply:
Or even it could be
```scala
val persons: List[Person] = ...

given byAge: Ordering[Person] = new Ordering[Person]:
  def compare(x: Person, y: Person): Int = x.age - y.age
  
persons.sorted(using byAge)
persons.sorted
```

Reply:
I pushed it

Reply:
I think this example is clearer, although I find it a little bit weird to sort people by age. We usually define type-class instances to provide a “default” behavior, or I would even say “THE obvious” behavior. In this case, I would find it non-idiomatic to go through a given instance to sort a list of people by age. Instead, I would just use `sortBy(_.age)`.

So, maybe we want to use another type than `Person`?

But, to say it again, I think the change you propose does not really show the value of implicits. People may just think “OK, it saves just passing one parameter, this is not really a big deal…”.

The goal of the original example was to show that the compiler can summon a complex ordering relationship, which would have been a pain to write manually (in this case, an ordering relationship based on two `String` criteria, so it compares the first criterion first, and falls back to the second criterion if the first does not discriminate the two values). IMHO this provides a higher value than what your example shows.

However, I can understand that maybe the way the original example was explained was not good enough and it may have seemed cryptic or maybe it failed to deliver its message?

Reply:
I agree that the original example shows a pattern that is much more useful and idiomatic than mine.

But there are so much things involved in it and kept hidden that it is really hard to understand. I think beginners would believe that the `sortBy` implementation does some magic (as if it was a macro or something) and they would miss the point and beauty of context parameters...

If we keep the original example I believe that we should show where the context is defined and how the parameter is summoned by the compiler and passed to the `sortBy` method.

Reply:
What about that, along with some explanations?

```scala
val addresses: List[Address] = ...

addresses.sortBy(address => (address.city, address.street))(Ordering.Tuple2(Ordering.String, Ordering.String))
addresses.sortBy(address => (address.city, address.street))
```

Reply:
Should I add a `import Ordering.given` to show where the context is defined?

Reply:
```
val addresses: List[Address] = ...

given ordering: Ordering[(String, String)] = Ordering.Tuple2(Ordering.String, Ordering.String)

addresses.sortBy(address => (address.city, address.street))(using ordering)
addresses.sortBy(address => (address.city, address.street))
```